### DISCUSSION TITLE
How do the transitions between hierarchies work?

### DISCUSSION THREAD

**Discussion by nicozacarias** [valhalla/valhalla#4762] [Q&A] [answered]
Hello,
I am trying to understand how the routing works and how the transition between different hierarchy levels occurs.
In the [hierarchy limits code](https://github.com/valhalla/valhalla/blob/master/valhalla/sif/hierarchylimits.h), I found these lines where default distances are defined:
```c++
// Default distances within which expansion is always allowed (per level). It's optimized
// for unidirectional search and can be modified by the path algorithm in case of
// bidirectional search.
constexpr float kDefaultExpansionWithinDist[] = {kMaxDistance, 100000.0f, 5000.0f, 0.0f,
                                                 0.0f,         0.0f,      0.0f,    0.0f};
}
``` 
Considering this GIF from the documentation as a reference, how does the path algorithm decide where to move upward or downward in levels?

![why_tiles](https://github.com/valhalla/valhalla/assets/48305269/54bd4718-6981-4dea-872b-a109c6269048)

How are the default distances used? At each step, does it measure the distance to the destination point and based on that decide whether to continue on the same level or not?

What happens if, while on a highway (level 0), it doesn't find another highway to continue on? Does it descend to a lower level, and then if the distance is still greater than 100km, ascend to another highway later on?

Thanks

**Comment by kevinkreiser** [comment, marked as answer]
yeah so i made that gif at the very beginning of the project and it doesnt reflect reality exactly. in practice mostly we use the bidirectional a* algorithm which sends out search paths from the start and end simultaneously. the hierarchy thresholds determine when each branch of the search decides to no longer explore a lower level of the hierarchy. so in practice near the source and destination you'll look at green tiles, as you get further away you'll stop lookign at them all together, until in the  middle you are only ever looking at the largest blue tiles.

we also have a fallback that if the route fails (you can imagine there is only a connection in the middle of a route on a lower hierarchy) we will run a second pass with relaxed limits for when different levels stop being used to hopefully let the route succeed.

**Reply by nicozacarias** [reply]
Ok now I got it, thank you for the quick and clear answer!
